rest-dashboard.go

```diff
@@ -308,6 +336,94 @@ func (r *Client) DeleteDashboard(ctx context.Context, slug string) (StatusMessag
 	return reply, err
 }
 
+type (
+	// SearchParam is a type for specifying SearchWithParams params.
+	SearchParam func(*url.Values)
+	// SearchType is a type accepted by WithSearchType func.
+	SearchType string
+)
+
+var (
+	SearchTypeFolder    SearchType = "dash-folder"
+	SearchTypeDashboard SearchType = "dash-db"
+)
+
+// WithSearchQuery specifies SearchWithParams search query.
+// Empty query is silently ignored.
+// Specifying it multiple times is futile, only last one will be sent.
+func WithSearchQuery(query string) SearchParam {
+	return func(v *url.Values) {
+		if query != "" {
+			v.Set("query", query)
+		}
+	}
+}
+
+// WithSearchTag specifies SearchWithParams tag to search for.
+// Empty tag is silently ignored.
+// Can be specified multiple times, logical OR is applied.
+func WithSearchTag(tag string) SearchParam {
+	return func(v *url.Values) {
+		if tag != "" {
+			v.Add("tag", tag)
+		}
+	}
+}
+
+// WithSearchType specifies SearchWithParams type to search for.
+// Specifying it multiple times is futile, only last one will be sent.
+func WithSearchType(searchType SearchType) SearchParam {
+	return func(v *url.Values) {
+		v.Set("type", string(searchType))
+	}
+}
+
+// WithSearchDashboardID specifies SearchWithParams dashboard id's to search for.
+// Can be specified multiple times, logical OR is applied.
+func WithSearchDashboardID(dashboardID int) SearchParam {
+	return func(v *url.Values) {
+		v.Add("dashboardIds", strconv.Itoa(dashboardID))
+	}
+}
+
+// WithSearchFolderID specifies SearchWithParams folder id's to search for.
+// Can be specified multiple times, logical OR is applied.
+func WithSearchFolderID(folderID int) SearchParam {
+	return func(v *url.Values) {
+		v.Add("folderIds", strconv.Itoa(folderID))
+	}
+}
+
+// WithSearchStarred specifies if SearchWithParams should search for starred dashboards only.
+// Specifying it multiple times is futile, only last one will be sent.
+func WithSearchStarred(starred bool) SearchParam {
+	return func(v *url.Values) {
+		v.Set("starred", strconv.FormatBool(starred))
+	}
+}
+
+// WithSearchLimit specifies maximum number of results from SearchWithParams query.
+// As of grafana 6.7 it has to be <= 5000. 0 stands for absence of parameter in a query.
+// Specifying it multiple times is futile, only last one will be sent.
+func WithSearchLimit(limit uint) SearchParam {
+	return func(v *url.Values) {
+		if limit > 0 {
+			v.Set("limit", strconv.FormatUint(uint64(limit), 10))
+		}
+	}
+}
+
+// WithSearchPage specifies SearchWithParams page number to be queried for.
+// Zero page is silently ignored, page numbers start from one.
+// Specifying it multiple times is futile, only last one will be sent.
+func WithSearchPage(page uint) SearchParam {
+	return func(v *url.Values) {
+		if page > 0 {
+			v.Set("page", strconv.FormatUint(uint64(page), 10))
+		}
```

Comment:
This is a tricky one - maybe it would make sense to `panic` here if `page == 0`? :thinking: I know that it's a bad practice to kill the programs of our users but without that someone might do `WithSearchPage(0)` which doesn't make sense. Not a hard request to change this - I wonder what's your opinion?

Reply:
Maybe not let's not panic, we would get a bunch of "bugs" :/

Reply:
I did some sanity check at first, but decided to get rid of it since it doesn't actually do much good for the user: grafana API _does_ accept `page=0` and just silently ignores it for now. I see no reason why we should enforce something grafana doesn't.

I strongly object to `panic` - instead we can easily implement a run-time check before actually sending a request:
```
if q.Get("page") == "0" { return _, err }
```

Should you like that, I'll add it. Just say if you think we need this check.